### PR TITLE
SDK easy buttons

### PR DESCRIPTION
**Pull request by nbrl**
Easy addition of Add To Any buttons by using an extra configuration object to `Amigo.init`. If no configuration is defined, the SDK works as it did before.

* Easy configuration and addition of buttons by Amigo configuration object.
* Configurable share channels.
* Size, colour and position can be modified.
* All Sharepoint widgets are configured together facilitating the use of the initialisation in *e.g.* a template file.
* Share URL and title can be configured per-Sharepoint.

Other changes:

* Changed `storeShareTokens` to store the whole share token object, but use `getShareTokens` to retrieve the simple array of tokens. This can be used later as a means to repopulate `Amigo.shareTokens` after a new page is loaded.
* Rewritten code to append or change the share token in a URL.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -225,6 +226,40 @@ Amigo.consumeToken('abc123', {
...
+            elementId: 'other-share-buttons'
+        }
+    }
+});
+```
+
+Note that you can declare the share widget configuration for all Sharepoints in one place, and only the widgets corresponding to a given location/URL will be rendered. When on a page matching `sharepointA`, but not `myOtherSharepoint`, the widgets for `sharepointA` will be loaded, and the configuration of `myOtherSharepoint` will be ignored.
+
+### Global configuration
+* `channels` *array* - an array containing Add To Any compatible share network strings. *Defaults to Facebook, Twitter and Email.*
```
Comment (jabclab) [inline README.md]: Worth linking to Add To Any's site?

File: gulpfile.js [hunk 2]
```diff
@@ -13,14 +13,18 @@ var sourcePaths = [
...
-var buildLibs = [
-  './bower_components/json2/json2.js',
-  './bower_components/history.js/scripts/bundled/html4+html5/native.history.js',
-  './bower_components/cookie/cookie.js',
-  './lib/cookieStorage.js', // requires cookie.js
-  './lib/localStorage.js',
-  './lib/sessionStorage.js',
-  './bower_components/fingerprint/fingerprint.js'
+var utilSrc = [
+    './bower_components/json2/json2.js',
```
Comment (jabclab) [inline gulpfile.js]: Was the indentation a bit off before when it was `buildLibs`?

File: src/sdk.js [hunk 3]
```diff
@@ -90,6 +91,17 @@
 
 
     /**
+     * Injection of other Amigo modules. Module "foo" for example,
+     * will be available at `Amigo.foo`.
+     * TODO: consider having this after the core is loaded.
```
Comment (jabclab) [inline src/sdk.js]: Is this `TODO` still applicable?

File: src/sdk.js [hunk 4]
```diff
@@ -124,45 +136,50 @@
 
         var url = window.location.href;
 
-        var qpRe = new RegExp('([?&]' + qpName + '=)([^&#]*)', 'g');
+        // Build a new share URL and just keep the query params onwards.
+        var newParams = '?' + Amigo._updateShareUrl(url, qpName, token).split('?')[1];
```
Comment (jabclab) [inline src/sdk.js]: How about changing `Amigo._updateShareUrl` so that it accepts a `url` of `undefined`, in which case it will only return the query string? Seems slightly nicer than building the full URL and then using `split('?')[1]`. What do you reckon?

File: src/sdk.js [hunk 5]
```diff
@@ -124,45 +136,50 @@
...
+     * @param {string} qpm - The query parameter key against which the token
+     *                       should be added/replaced.
+     * @param {string} token - The token to add/replace to/in the URL.
+     * @returns {string} - URL containing the added/replaced token.
+     */
+    Amigo._updateShareUrl = function (url, qpm, token) {
+        // url may or may not contain a token already so need to strip it out first.
+        var re = new RegExp('([?&]' + qpm + '=)([0-9a-zA-Z]+)');
+        var separator = (url.indexOf('?') !== -1)
+                ? '&'
```
Comment (jabclab) [inline src/sdk.js]: If we're using multi-line ternary do we want the `?` and `:` to line up with the `=`? e.g.

```js
var separator = (url.indexOf('?') !== -1)
                      ? '&'
                      : '?';
```

Don't mind too much. What do you reckon?

File: src/sdk.js [hunk 6]
```diff
@@ -201,7 +218,7 @@
      * @returns {array} - Array of strings of sharepoint tokens from storage.
      * // TODO: fix
      */
-    Amigo._getTouchpointShareTokens = function () {
+    Amigo._getReferralShareTokens = function () {
```
Comment (jabclab) [inline src/sdk.js]: Do we want to go down as far as "referral"? How about `Amigo._getTouchShareTokens`?

File: src/sdk.js [hunk 7]
```diff
@@ -201,7 +218,7 @@
      * @returns {array} - Array of strings of sharepoint tokens from storage.
      * // TODO: fix
```
Comment (jabclab) [inline src/sdk.js]: Is this `TODO` still applicable?

File: src/sdk.js [hunk 8]
```diff
@@ -926,11 +971,47 @@
     };
 
     /**
-     * Retrieve the default token, at present the first.
-     * @returns {string} - The default token for the page.
+     * Retrieve the token (if any) from the provided URL.
+     * @param {string} url - The URL with a token.
+     * @param {string} qpm - The key of the query parameter in the url.
```
Comment (jabclab) [inline src/sdk.js]: Think this should be identified as optional.

File: src/sdk.js [hunk 9]
```diff
@@ -926,11 +971,47 @@
...
+
+    /**
+     * Given a recent token, will return the name of the sharepoint with which
+     * it is associated.
+     * @param {string} token - The token for whom the associated sharepoint name
+     *                         is desired.
+     * @returns {string} - The name of the associated sharepoint, if found.
+     */
+    Amigo.getSharepointNameFromToken = function (token) {
+        var sharepoint = null;
```
Comment (jabclab) [inline src/sdk.js]: Could declare after the `if (!Amigo.shareTokens)`.

File: src/sdk.js [hunk 10]
```diff
@@ -926,11 +971,47 @@
...
+            return null;
+        }
+
+        for (var i=0,len=Amigo.shareTokens.length; i<len; i++) {
+            if (token === Amigo.shareTokens[i].token) {
+                sharepoint = Amigo.shareTokens[i].name;
+            }
+        }
+
+        return sharepoint || null;
```
Comment (jabclab) [inline src/sdk.js]: Is the ` || null` required?

File: src/buttons.js [hunk 11]
```diff
@@ -0,0 +1,238 @@
...
+        }
+    };
+
+    /**
+     * Globally sets the background colour for Add To Any buttons.
+     * @param {string} - Hex colour (including hash) or HTML colour name.
+     */
+    Buttons._setButtonColour = function (colour) {
+        window.a2a_config = window.a2a_config || {};
+        window.a2a_config.icon_color = colour;
```
Comment (jabclab) [inline src/buttons.js]: If in the (albeit unlikely) circumstances that the client already has a2a buttons would this mean all the other buttons change colour?

File: src/buttons.js [hunk 12]
```diff
@@ -0,0 +1,238 @@
...
+     * @param {array} sharepointMeta - Array of Sharepoint metadata describing
+     *                                 the Sharepoints and tokens for the current
+     *                                 page.
+     */
+    Buttons._addOrUpdateShareButtons = function (sharepointMeta) {
+        Amigo._log('info', 'Buttons._addOrUpdateShareButtons()');
+        var targets = config.buttons.sharepoints;
+        for (var i=0,len=sharepointMeta.length; i<len; i++) {
+            if (targets[sharepointMeta[i].name] && targets[sharepointMeta[i].name].elementId) {
+                if (document.getElementsByClassName('amigo-button-' + sharepointMeta[i].name).length === 0) {
```
Comment (jabclab) [inline src/buttons.js]: Is this safe given that the sharepoint name might contain internal whitespace? e.g.

```js
document.getElementsByClassName('amigo-button-My Long Sharepoint Name')
```

File: src/buttons.js [hunk 13]
```diff
@@ -0,0 +1,238 @@
...
+    };
+
+    /**
+     * Initialise the Buttons module, if needed.
+     */
+    Buttons._init = function () {
+        Amigo._log('info', 'Buttons._init()');
+
+        // Load Add to Any JS.
+        Buttons._loadAddToAny();
```
Comment (jabclab) [inline src/buttons.js]: Do we need to synchronise this? e.g. could the buttons load before the configuration has been set?

File: README.md (hunk 1 above)
Comment (nbrl) [inline README.md]: Done.

File: gulpfile.js (hunk 2 above)
Comment (nbrl) [inline gulpfile.js]: Yeah, the `gulpfile` was 2 space indented for some reason, and that must have been left over from getting it to 4 space - now the whole file is 4, to match the rest of the repo.

File: src/sdk.js (hunk 3 above)
Comment (nbrl) [inline src/sdk.js]: Nope, removed.

File: src/buttons.js [hunk 14]
```diff
@@ -0,0 +1,238 @@
...
+     */
+    Buttons._addShareButtonGroupToDom = function (location, sharepoint, channels, size, shareUrl, shareTitle) {
+        if (document.readyState === 'interactive' || document.readyState === 'complete') {
+            var buttonLocation = document.getElementById(location);
+            if (buttonLocation) {
+                buttonLocation.appendChild(Buttons._buildShareButtonGroup(sharepoint, channels, size));
+                Buttons._updateButtonConfig(sharepoint, shareUrl, shareTitle);
+            }
+        } else {
+            setTimeout(arguments.callee, 100);
```
Comment (jabclab) [inline src/buttons.js]: `arguments.callee` [isn't permitted in strict mode](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/arguments/callee). How about using one of MDN's suggestions here?

File: src/buttons.js [hunk 15]
```diff
@@ -0,0 +1,238 @@
...
+        var attributes = [
+            'a2a_kit',
+            'a2a_kit_size_' + size,
+            'a2a_default_style',
+            'a2a_target',
+            'amigo-button-' + sharepoint
+        ].join(' ');
+
+        var buttonEl = document.createElement('div');
+        buttonEl.className = attributes; // TODO: check IE
```
Comment (jabclab) [inline src/buttons.js]: `TODO` can be removed now?

File: src/buttons.js [hunk 16]
```diff
@@ -0,0 +1,238 @@
...
+        // Assign defaults if necessary
+        channels = channels || DEFAULT_SHARE_CHANNELS;
+        size = BUTTON_SIZES[size] || DEFAULT_BUTTON_SIZE;
+
+        var attributes = [
+            'a2a_kit',
+            'a2a_kit_size_' + size,
+            'a2a_default_style',
+            'a2a_target',
+            'amigo-button-' + sharepoint
```
Comment (jabclab) [inline src/buttons.js]: Same issue with the sharepoint names containing whitespace as mentioned below

File: src/buttons.js [hunk 17]
```diff
@@ -0,0 +1,238 @@
...
+     * Used after a new token is generated to replace the share token in the
+     * desired button share URLs.
+     * @param {string} sharepoint - The name of the Sharepoint which the buttons
+     *                              are associated with.
+     * @param {string} shareUrl - The base URL that the buttons will be sharing.
+     * @param {string} shareTitle -  The title that will be the label for the share.
+     */
+    function updateButton (sharepoint, shareUrl, shareTitle) {
+        a2a.init('page', {
+            target: '.amigo-button-' + sharepoint,
```
Comment (jabclab) [inline src/buttons.js]: Sharepoint containing whitespace again :smiley_cat:

File: src/buttons.js [hunk 18]
```diff
@@ -0,0 +1,238 @@
...
+        var targets = config.buttons.sharepoints;
+        for (var i=0,len=sharepointMeta.length; i<len; i++) {
+            if (targets[sharepointMeta[i].name] && targets[sharepointMeta[i].name].elementId) {
+                if (document.getElementsByClassName('amigo-button-' + sharepointMeta[i].name).length === 0) {
+                    Buttons._addShareButtonGroupToDom(
+                        targets[sharepointMeta[i].name].elementId,
+                        sharepointMeta[i].name,
+                        config.buttons.channels,
+                        config.buttons.size,
+                        Amigo._updateShareUrl(
```
Comment (jabclab) [inline src/buttons.js]: This is used here and in the `Buttons._updateButtonConfig` below. Could we declare it once? Also, as `Buttons._addShareButtonGroupToDom` also does the update would we want to do it something like:

```
if (document.getElementsByClassName('amigo-button-' + sharepointMeta[i].name).length === 0) {
  Buttons._addShareButtonGroupToDom(/* ... */);
}

Buttons._updateButtonConfig(/* ... */)
```

That would allow us to remove the update from `Buttons._addShareButtonGroupToDom`. What do you think?

File: src/sdk.js (hunk 4 above)
Comment (nbrl) [inline src/sdk.js]: Yep good call, done.

File: src/sdk.js (hunk 5 above)
Comment (nbrl) [inline src/sdk.js]: We do - this is emacs having a very strong opinion about where its tabstop is. Done manually for now.

File: src/sdk.js (hunk 6 above)
Comment (nbrl) [inline src/sdk.js]: Strictly speaking it will only be referralTokens as if a touch is not a referral, the token won't be stored.

Still, can change to touch - what do you think?

File: src/sdk.js (hunk 7 above)
Comment (nbrl) [inline src/sdk.js]: Nope. Removed.

File: src/sdk.js (hunk 8 above)
Comment (nbrl) [inline src/sdk.js]: Done.

File: src/sdk.js (hunk 9 above)
Comment (nbrl) [inline src/sdk.js]: Done.

File: src/sdk.js (hunk 10 above)
Comment (nbrl) [inline src/sdk.js]: Nope, just testing you.

**Comment by jabclab** [comment]
Have had a play on [JSFiddle](http://jsfiddle.net/jakeclarkson/1htywmj3/) with the sharepoint name containing whitespace. It appears to work but, as the whitespace effectively means a new class, we could inadvertently inherit an existing CSS class (as is the case in my example).

File: src/buttons.js (hunk 11 above)
Comment (nbrl) [inline src/buttons.js]: Yep, we'd also be loading the A2A JS twice and adding meaningless (for the non-Amigo A2A buttons) callbacks. In this case though, it looks like we could set it as a `data_a2a_` attribute on the `div`?

File: src/buttons.js (hunk 12 above)
Comment (nbrl) [inline src/buttons.js]: Fair point. It will work, but it isn't pretty as each space is a new class. E.g. the above example has four classes: `amigo-button-My`, `Long`, `Sharepoint` and `Name`. Just trying to work out why I didn't use an ID...

File: src/buttons.js (hunk 13 above)
Comment (nbrl) [inline src/buttons.js]: Doesn't matter. The HTML can go in first, at which point the initialisation is done via registering an A2A callback, or if the JS has loaded first, the initialisation will be called directly after adding the HTML.

File: src/buttons.js (hunk 14 above)
Comment (nbrl) [inline src/buttons.js]: Will have a look, but since we're not using strict it shouldn't matter.

File: src/buttons.js (hunk 15 above)
Comment (nbrl) [inline src/buttons.js]: Yep. Done.

File: src/buttons.js (hunk 14 above)
Comment (nbrl) [inline src/buttons.js]: Just call the function! smh

File: gulpfile.js (hunk 2 above)
Comment (jabclab) [inline gulpfile.js]: Cool :smile:

File: src/buttons.js [hunk 19]
```diff
@@ -0,0 +1,229 @@
...
+     * @param {array} channels - An array of channels used, corresponding to the
+     *                           desired buttons.
+     * @param {string} size - Button size name, S, M or L.
+     */
+    Buttons._addShareButtonGroupToDom = function (location, sharepoint, channels, size, colour) {
+        if (document.readyState === 'interactive' || document.readyState === 'complete') {
+            var buttonLocation = document.getElementById(location);
+            if (buttonLocation) {
+                buttonLocation.appendChild(Buttons._buildShareButtonGroup(sharepoint, channels, size, colour));
+            }
```
Comment (jabclab) [inline src/buttons.js]: Worth logging a warning if the specified element does not exist?

File: src/sdk.js [hunk 20]
```diff
@@ -468,16 +485,35 @@
...
+                Amigo.shareTokens.push(data[i]);
+            }
+        }
+
         // Retrieve what is currently in storage
         var current = JSON.parse(sessionStore.getItem(SESSION_STORAGE_NAME));
 
         var apiKey = config.apiKey;
 
+        // TODO: Relying on Amigo.shareTokens means we won't have any stored on
```
Comment (jabclab) [inline src/sdk.js]: `TODO` still applicable?

File: src/sdk.js [hunk 21]
```diff
@@ -934,6 +979,50 @@
...
+    Amigo.getSharepointNameFromToken = function (token) {
+        if (!Amigo.shareTokens) {
+            return null;
+        }
+
+        var sharepoint = null;
+
+        for (var i=0,len=Amigo.shareTokens.length; i<len; i++) {
+            if (token === Amigo.shareTokens[i].token) {
+                sharepoint = Amigo.shareTokens[i].name;
```
Comment (jabclab) [inline src/sdk.js]: Fine to merge but could this just be `return Amigo.shareTokens[i].name;`?

File: src/buttons.js (hunk 13 above)
Comment (jabclab) [inline src/buttons.js]: Cool, :+1:

File: src/buttons.js [hunk 22]
```diff
@@ -0,0 +1,229 @@
...
+     * Used after a new token is generated to replace the share token in the
+     * desired button share URLs.
+     * @param {string} sharepoint - The name of the Sharepoint which the buttons
+     *                              are associated with.
+     * @param {string} shareUrl - The base URL that the buttons will be sharing.
+     * @param {string} shareTitle -  The title that will be the label for the share.
+     */
+    function updateButton (sharepoint, shareUrl, shareTitle) {
+        window.a2a.init('page', {
+            target: '#' + AMIGO_BUTTON_GROUP_PREFIX + sharepoint,
```
Comment (jabclab) [inline src/buttons.js]: http://stackoverflow.com/questions/1391802/can-a-dom-element-have-an-id-that-contains-a-space may be worth a read

File: src/buttons.js [hunk 23]
```diff
@@ -0,0 +1,229 @@
...
+
+    /**
+     * Loads the Add To Any JS asynchronously if it is not already present.
+     */
+    Buttons._loadAddToAny = function () {
+        if (!window.a2a) {
+            var a = document.createElement('script');
+            a.type = 'text/javascript';
+            a.async = true;
+            a.src = '//static.addtoany.com/menu/page.js';
```
Comment (jabclab) [inline src/buttons.js]: Does Add To Any provide versioned scripts?

File: test/unit/sdk/_storeShareTokens.js [hunk 24]
```diff
@@ -0,0 +1,208 @@
+var expect = require('expect.js');
+var sinon = require('sinon');
+
+var apiKey = 'foobarbaz';
+var defaultSessionStorageName = 'amigo-session';
+
+describe('_storeShareTokens()', function () {
+
+    beforeEach(function () {
+	sinon.sandbox.create();
```
Comment (jabclab) [inline test/unit/sdk/_storeShareTokens.js]: Wonky indentation?

File: test/unit/sdk/_storeShareTokens.js [hunk 25]
```diff
@@ -0,0 +1,208 @@
...
+    });
+
+    it('should have a _storeShareTokens function', function () {
+        expect(Amigo._storeShareTokens).to.be.a('function');
+    });
+
+    it('should return immediately if the passed data is not an array', function () {
+        var jsonParseSpy = sinon.sandbox.spy(window.JSON, 'parse');
+
+	Amigo._storeShareTokens(null);
```
Comment (jabclab) [inline test/unit/sdk/_storeShareTokens.js]: Indentation

File: src/buttons.js (hunk 23 above)
Comment (nbrl) [inline src/buttons.js]: Nope.

File: test/unit/buttons/_buildShareButtonGroup.js [hunk 26]
```diff
@@ -0,0 +1,152 @@
...
+        // Act
+        var actual = Amigo.Buttons._buildShareButtonGroup(sharepoint, channels, size);
+
+        // Assert
+        expect(actual.outerHTML).to.equal(expected.outerHTML);
+    });
+
+    it('should fallback to a medium size if a nonsensical size is provided', function () {
+        // Arrange
+	var sharepoint = 'sp1';
```
Comment (jabclab) [inline test/unit/buttons/_buildShareButtonGroup.js]: Indentation

File: test/unit/buttons/_buildShareButton.js [hunk 27]
```diff
@@ -0,0 +1,45 @@
...
+    });
+
+    it('should return a button element with the correct class for the channel', function () {
+        // Arrange
+        var channel = 'facebook';
+        var expected = document.createElement('a');
+        expected.className = 'a2a_button_' + channel;
+
+        // Act
+	var actual = Amigo.Buttons._buildShareButton(channel);
```
Comment (jabclab) [inline test/unit/buttons/_buildShareButton.js]: Indentation

File: src/buttons.js (hunk 22 above)
Comment (nbrl) [inline src/buttons.js]: That's annoying - either we restrict sharepoint names, or build an internal map of sharepoint name to whatever we call the id. Probably the latter, but it isn't going to get done today, dragging this on.

Comment (jabclab) [inline src/buttons.js]: Maybe just add a test to see if a sharepoint containing whitespace works in all our target browsers?

File: src/buttons.js (hunk 23 above)
Comment (jabclab) [inline src/buttons.js]: :+1:

File: src/buttons.js (hunk 19 above)
Comment (nbrl) [inline src/buttons.js]: Yep.

File: src/sdk.js (hunk 20 above)
Comment (nbrl) [inline src/sdk.js]: Yep. Basically saying we set up `Amigo.shareTokens` then copy it to `sessionStorage`, but in the future we'll likely want to copy out of `sessionStorage` back into `Amigo.shareTokens` for use on subsequent pages.

File: src/sdk.js (hunk 21 above)
Comment (nbrl) [inline src/sdk.js]: Yep.

File: src/buttons.js (hunk 22 above)
Comment (nbrl) [inline src/buttons.js]: OK. I will add the above somewhere as well though (issue or LP) as it should probably be standards compliant.

File: test/unit/sdk/_storeShareTokens.js (hunk 24 above)
Comment (nbrl) [inline test/unit/sdk/_storeShareTokens.js]: Where?

Comment (nbrl) [inline test/unit/sdk/_storeShareTokens.js]: Ah, sneaky tab.

File: test/unit/sdk/_storeShareTokens.js (hunk 25 above)
Comment (nbrl) [inline test/unit/sdk/_storeShareTokens.js]: `untabify`-d.

File: test/unit/buttons/_buildShareButtonGroup.js (hunk 26 above)
Comment (nbrl) [inline test/unit/buttons/_buildShareButtonGroup.js]: And again.

File: test/unit/buttons/_buildShareButton.js (hunk 27 above)
Comment (nbrl) [inline test/unit/buttons/_buildShareButton.js]: Not entirely sure where the tabs are coming from...

File: test/unit/buttons/buttons.js [hunk 28]
```diff
@@ -0,0 +1,41 @@
...
+        });
+    });
+
+    afterEach(function () {
+        sinon.sandbox.restore();
+
+        this.xhr.restore();
+    });
+
+    it.only('should correctly identify an elements with an ID with spaces', function () {
```
Comment (jabclab) [inline test/unit/buttons/buttons.js]: Remove the `.only`

File: src/buttons.js [hunk 29]
```diff
@@ -0,0 +1,231 @@
...
+
+    /**
+     * Adds a particular share button group to the correct location in the DOM.
+     * @param {string} location - An HTML element ID in which the buttons should
+     *                            be placed.
+     * @param {string} sharepoint - The name of the Sharepoint which the buttons
+     *                              are associated with.
+     * @param {array} channels - An array of channels used, corresponding to the
+     *                           desired buttons.
+     * @param {string} size - Button size name, S, M or L.
```
Comment (jabclab) [inline src/buttons.js]: Missing `colour` in the JSDoc.

File: src/buttons.js [hunk 30]
```diff
@@ -0,0 +1,231 @@
...
+    Buttons._addShareButtonGroupToDom = function (location, sharepoint, channels, size, colour) {
+        if (document.readyState === 'interactive' || document.readyState === 'complete') {
+            var buttonLocation = document.getElementById(location);
+            if (buttonLocation) {
+                buttonLocation.appendChild(Buttons._buildShareButtonGroup(sharepoint, channels, size, colour));
+            } else {
+                Amigo._log('warn', 'Requested element does not exist. ID = ' + buttonLocation);
+            }
+        } else {
+            setTimeout(Buttons._addShareButtonGroupToDom.apply(null, arguments), 100);
```
Comment (jabclab) [inline src/buttons.js]: Might want to prefix `setTimeout` with `window`
